activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb

```diff
@@ -558,21 +558,27 @@ def remove(conn)
       # or a thread dies unexpectedly.
       def reap
         stale_connections = synchronize do
-          @connections.select do |conn|
-            conn.in_use? && !conn.owner.alive?
-          end.each do |conn|
-            conn.steal!
-          end
+          @connections.select(&:stale?).each(&:steal!)
+        end
+
+        exited_connections = synchronize do
+          @connections.select(&:exited?).each(&:steal!)
         end
 
-        stale_connections.each do |conn|
-          if conn.active?
-            conn.reset!
-            checkin conn
+        return if stale_connections.empty? && exited_connections.empty?
```

Comment:
Drive by review: don't think this is needed, since an empty array won't execute the body of an `each`.